### ISSUE TITLE
Kubectl/API Refusal to delete ReplicaSet

### ISSUE THREAD

**Issue by mccormd** [kubernetes/kubernetes#47960]
/kind bug
**What happened**:
Tried to delete a replicaset with a failing container (stuck on a volume mount) and got the following: -

~~~
kubectl delete rs prometheus-4088412010 -n kube-system 
error: Scaling the resource failed with: ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified; Current resource version 11276377
~~~

Looking at the resource looks odd: -
~~~
# kubectl get rs prometheus-4088412010 -n kube-system -o yaml
apiVersion: extensions/v1beta1
kind: ReplicaSet
metadata:
  annotations:
    deployment.kubernetes.io/desired-replicas: "1"
    deployment.kubernetes.io/max-replicas: "2"
    deployment.kubernetes.io/revision: "2"
  creationTimestamp: 2017-06-23T10:10:25Z
  deletionGracePeriodSeconds: 0
  deletionTimestamp: 2017-06-23T12:00:39Z
  finalizers:
  - foregroundDeletion
  generation: 2
  labels:
    app: prometheus
    pod-template-hash: "4088412010"
  name: prometheus-4088412010
  namespace: kube-system
  ownerReferences:
  - apiVersion: extensions/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: Deployment
    name: prometheus
    uid: a8870531-5297-11e7-a80e-0a80992c354e
  resourceVersion: "11276377"
  selfLink: /apis/extensions/v1beta1/namespaces/kube-system/replicasets/prometheus-4088412010
  uid: 2ccd6bf2-57fc-11e7-bc44-0a851b7f13b8
spec:
  replicas: 1
  selector:
    matchLabels:
      app: prometheus
      pod-template-hash: "4088412010"
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: prometheus
        pod-template-hash: "4088412010"
      name: prometheus
    spec:
      containers:
      - args:
        - -storage.local.retention=360h0m0s
        - -storage.local.memory-chunks=500000
        - -config.file=/etc/prometheus/prometheus.yml
        image: quay.io/prometheus/prometheus:v1.7.1
        imagePullPolicy: IfNotPresent
        name: prometheus
        ports:
        - containerPort: 9090
          name: web
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/prometheus
          name: config-volume
        - mountPath: /prometheus
          name: data-volume
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: prometheus
      serviceAccountName: prometheus
      terminationGracePeriodSeconds: 30
      volumes:
      - configMap:
          defaultMode: 420
          name: prometheus
        name: config-volume
      - name: data-volume
        persistentVolumeClaim:
          claimName: k8smetrics-prometheus
status:
  fullyLabeledReplicas: 1
  observedGeneration: 1
  replicas: 1
~~~

~~~
# kubectl describe rs prometheus-4088412010 -n kube-system        
Name:		prometheus-4088412010
Namespace:	kube-system
Selector:	app=prometheus,pod-template-hash=4088412010
Labels:		app=prometheus
		pod-template-hash=4088412010
Annotations:	deployment.kubernetes.io/desired-replicas=1
		deployment.kubernetes.io/max-replicas=2
		deployment.kubernetes.io/revision=2
Replicas:	1 current / 1 desired
Pods Status:	1 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:		app=prometheus
			pod-template-hash=4088412010
  Service Account:	prometheus
  Containers:
   prometheus:
    Image:	quay.io/prometheus/prometheus:v1.7.1
    Port:	9090/TCP
    Args:
      -storage.local.retention=360h0m0s
      -storage.local.memory-chunks=500000
      -config.file=/etc/prometheus/prometheus.yml
    Environment:	<none>
    Mounts:
      /etc/prometheus from config-volume (rw)
      /prometheus from data-volume (rw)
  Volumes:
   config-volume:
    Type:	ConfigMap (a volume populated by a ConfigMap)
    Name:	prometheus
    Optional:	false
   data-volume:
    Type:	PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:	k8smetrics-prometheus
    ReadOnly:	false
Events:		<none>
~~~

**What you expected to happen**:
I was expecting the ReplicaSet to be deleted.

**How to reproduce it (as minimally and precisely as possible)**:

Created with this deployment: -
~~~
apiVersion: v1
kind: Service
metadata:
  annotations:
    prometheus.io/scrape: 'true'
  namespace: kube-system
  labels:
    name: prometheus
    addonmanager.kubernetes.io/mode: EnsureExists
  name: prometheus
spec:
  selector:
    app: prometheus
  type: ClusterIP
  ports:
  - name: prometheus
    protocol: TCP
    port: 9090
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: prometheus
  namespace: kube-system
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
spec:
  replicas: 1
  selector:
    matchLabels:
      app: prometheus
  template:
    metadata:
      name: prometheus
      labels:
        app: prometheus
    spec:
      serviceAccountName: prometheus
      containers:
      - name: prometheus
        image: quay.io/prometheus/prometheus:v1.7.1
        args:
          - '-storage.local.retention=360h0m0s'
          - '-storage.local.memory-chunks=500000'
          - '-config.file=/etc/prometheus/prometheus.yml'
        ports:
        - name: web
          containerPort: 9090
        volumeMounts:
        - name: config-volume
          mountPath: /etc/prometheus
        - name: data-volume
          mountPath: /prometheus
      volumes:
      - name: config-volume
        configMap:
          name: prometheus
      - name: data-volume
        persistentVolumeClaim:
          claimName: k8smetrics-prometheus
~~~

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"clean", BuildDate:"2017-05-19T18:33:17Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"clean", BuildDate:"2017-05-19T18:33:17Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration**: aws
- OS (e.g. from /etc/os-release):
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1409.5.0
VERSION_ID=1409.5.0
BUILD_ID=2017-06-22-2222
PRETTY_NAME="Container Linux by CoreOS 1409.5.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"

- Kernel (e.g. `uname -a`): 
Linux ip-10-16-5-136.eu-west-1.compute.internal 4.11.6-coreos-r1 #1 SMP Thu Jun 22 22:04:38 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz GenuineIntel GNU/Linux
- Install tools: roll-your-own
- Others:

**Comment by mccormd** [comment]
I see these logs in the controller: -

~~~
I0623 12:00:03.168839       1 graph_builder.go:391] add [extensions/v1beta1/ReplicaSet, namespace: kube-system, name: prometheus-4088412010, uid: 2ccd6bf2-57fc-11e7-bc44-0a851b7f13b8] to the attemptToDelete, because it's waiting for its dependents to be deleted
I0623 12:00:03.168895       1 garbagecollector.go:271] processing item [v1/Pod, namespace: kube-system, name: prometheus-4088412010-bk19v, uid: 2cd0250c-57fc-11e7-bc44-0a851b7f13b8]
I0623 12:00:03.169191       1 garbagecollector.go:271] processing item [extensions/v1beta1/ReplicaSet, namespace: kube-system, name: prometheus-4088412010, uid: 2ccd6bf2-57fc-11e7-bc44-0a851b7f13b8]
I0623 12:00:03.174893       1 garbagecollector.go:376] adding [v1/Pod, namespace: kube-system, name: prometheus-4088412010-bk19v, uid: 2cd0250c-57fc-11e7-bc44-0a851b7f13b8] to attemptToDelete, because its owner [extensions/v1beta1/ReplicaSet, namespace: kube-system, name: prometheus-4088412010, uid: 2ccd6bf2-57fc-11e7-bc44-0a851b7f13b8] is deletingDependents
E0623 12:00:03.179180       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
I0623 12:00:03.180251       1 garbagecollector.go:362] delete object [v1/Pod, namespace: kube-system, name: prometheus-4088412010-bk19v, uid: 2cd0250c-57fc-11e7-bc44-0a851b7f13b8] with Default
I0623 12:00:03.238811       1 garbagecollector.go:271] processing item [v1/Pod, namespace: kube-system, name: prometheus-4088412010-bk19v, uid: 2cd0250c-57fc-11e7-bc44-0a851b7f13b8]
E0623 12:00:03.240165       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:03.251223       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:03.267759       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:03.313756       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:03.399520       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:03.565355       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:03.892543       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:04.537915       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:05.824190       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
E0623 12:00:08.389849       1 replica_set.go:424] Sync "kube-system/prometheus-4088412010" failed with ReplicaSet.extensions "prometheus-4088412010" is invalid: metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
~~~

**Comment by dchen1107** [comment]
@kubernetes/sig-api-machinery-bugs @kubernetes/sig-api-machinery-misc Can someone in sig-api-machinery decide if this is a release blocker for 1.7? The issue is found in 1.6.4. is there any workaround for the users?

**Comment by caesarxuchao** [comment]
Should be fixed by https://github.com/kubernetes/kubernetes/pull/46500. Might need cherry-pick to 1.6

**Comment by caesarxuchao** [comment]
The rs controller tried to sync a rs that has the foregroundGC finalizers, and triggered the bug.

It's fixed in 1.7, we should backport it to 1.6.

**Comment by caesarxuchao** [comment]
Sent https://github.com/kubernetes/kubernetes/pull/47994 to cherrypick to 1.6.

**Comment by enisoc** [comment]
@caesarxuchao Why doesn't this happen for every `kubectl delete rs`? Is it a race? Or dependent on `kubectl` version?

**Comment by caesarxuchao** [comment]
@mccormd do you know who put the "foregroundDeletion" finalizer on the RS? 

Is the cluster downgraded from 1.7 to 1.6? Because i know in 1.7 the deployment controller adds the "foregroundDeletion" finalizer for the RS, but 1.6 shouldn't have this behavior.

**Comment by mccormd** [comment]
@caesarxuchao I'm not aware on how the "forgroundDeletion" finalizer gets on there - I don't understand what it is even.  I just deploy using the yaml manifest I pasted above which is picked up by the addon manager.

**Comment by caesarxuchao** [comment]
I don't know how a 1.6 rs ends up with the "foregroundDeletion" finalizer. The bug should be fixed by #47994. Closing the issue.

**Comment by jacohend** [comment]
Still encountering this issue

**Comment by enisoc** [comment]
@jacohend Can you share the output of `kubectl version`? A small example manifest that triggers the problem would also be helpful. Thanks.

**Comment by hmarcelodn** [comment]
@enisoc we are experiencing the same issue at production suddenly. Is there any real solution for this?

**Comment by hmarcelodn** [comment]
@enisoc 

Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.3", GitCommit:"2c2fe6e8278a5db2d15a013987b53968c743f2a1", GitTreeState:"clean", BuildDate:"2017-08-03T07:00:21Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.6", GitCommit:"7fa1c1756d8bc963f1a389f4a6937dc71f08ada2", GitTreeState:"clean", BuildDate:"2017-06-16T18:21:54Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}

**Comment by enisoc** [comment]
@hmarcelodn If your symptom is `Invalid value: "foregroundDeletion"`, it should be fixed if you upgrade the cluster to v1.6.7 or higher.
